typeform/typeform.py

```diff
@@ -0,0 +1,183 @@
+DESTINATION_POSTFIX = '_{__table}'
+BASE_URL = 'https://api.typeform.com/v1'
+DATE_PARSER_FORMAT = '%Y-%m-%dT%H:%M:%S'
+
+
+class Typeform(panoply.DataSource):
+
+    # constructor
+    def __init__(self, source, options):
+        super(Typeform, self).__init__(source, options)
+
+        if 'destination' not in source:
+            source['destination'] = DESTINATION
+
+        # append the destination postfix, which represent the form name
+        source['destination'] += DESTINATION_POSTFIX
+
+        # configure incremental pulling
+        self._incval = None
+        if (source.get('inckey')):
+            incval = source.get('incval')
+            self._incval = (incval if incval else
+                source.get('lastTimeSucceed') or source.get('lastRuntime'))
+
+            # validate that's the incremental value is a valid
+            # date that could be parsed to a Unix timestamp before
+            # including it in the process
+            self._incval = getTimestamp(self._incval)
+
+
+        forms = source.get('forms')
+        forms = forms if forms else []
+
+        # add an 'offset' attribute used for pagination for each
+        # different form, used as the actual number to use as offset
+        self._forms = map(lambda form: {
+            'id': form['id'],
+            'name': form['name'],
+            'offset': 0
+        }, forms)
+
+        self._key = source.get('key')
+        self._total = len(self._forms)
+
+    def read(self, n = None):
+        if len(self._forms) == 0:
+            # no more data to consume
+            return None
+
+        form = self._forms[0]
+        url = '%s/form/%s?key=%s&completed=true&offset=%s&limit=%s' % (
+            BASE_URL,
+            form['id'],
+            self._key,
+            form['offset'],
+            FETCH_LIMIT
+        )
+
+        # pull data incrementally if configured to do so.
+        if self._incval:
+            url += '&since=%s' % (self._incval) 
+
+        body = self._request(url)
+
+        # 'completed' represent the number of completed forms that
+        # returned from the server
+        stats = body.get('stats').get('responses')
+        total = stats['completed']
+
+        if (form['offset'] + FETCH_LIMIT) >= total:
+            # we're done paginating with the current form.
+            form['offset'] = None
+
+            # no more results for this form, remove it
+            self._forms.pop(0)
+
+            # report progress
+            loaded = self._total - len(self._forms)
+            msg = '%s of %s forms fetched' % (loaded, self._total)
+            self.progress(loaded, self._total, msg)
+        else:
+            # prepare the offset to the next set of records
+            form['offset'] += FETCH_LIMIT
+
+
+        def add_dest(dest):
+            def add(item):
+                item['__table'] = dest
+                return item
+
+            return add
+
+        # clean up irrelevant attributes and add the
```

Comment:
What cleanup?
are you talking about the fact that you only take `questions` and `responses` from the `body`?